File: packages/agent/src/context/context-manager.ts
/**
 * @fileoverview Context Manager
 *
 * Central management of all context for agent calls:
 * - Message tracking with token estimation
 * - Model-aware context limits
 * - Pre-turn validation
 * - Compaction orchestration
 * - Tool result processing
 * - Model switching with revalidation
 */

import type { Message, Tool } from '@core/types/index.js';
import { getContextLimit } from '@infrastructure/usage/index.js';
import { detectProviderFromModel, type ProviderType } from '@llm/providers/index.js';
import type { Summarizer } from './summarizer.js';
import {
  buildSystemPrompt,
  requiresToolClarificationMessage,
  getToolClarificationMessage,
  TRON_CORE_PROMPT,
  loadSystemPromptFromFileSync,
} from './system-prompts.js';
import {
  estimateMessageTokens as estimateMessageTokensUtil,
  estimateRulesTokens as estimateRulesTokensUtil,
  estimateSystemPromptTokens as estimateSystemPromptTokensUtil,
  estimateToolsTokens as estimateToolsTokensUtil,
} from './token-estimator.js';
import { MessageStore } from './message-store.js';
import { CompactionEngine } from './compaction-engine.js';
import { ContextSnapshotBuilder } from './context-snapshot-builder.js';
import {
  type ThresholdLevel,
  type ContextManagerConfig,
  type ContextSnapshot,
  type DetailedContextSnapshot,
  type PreTurnValidation,
  type CompactionPreview,
  type CompactionResult,
  type ProcessedToolResult,
  type ExportedState,
} from './types.js';

// =============================================================================
// ContextManager
// =============================================================================

/**
 * Central context management for agent calls.
 *
 * Responsibilities:
 * - Track messages and estimate tokens
 * - Provide model-aware context limits
 * - Pre-validate turns before API calls
 * - Orchestrate compaction
 * - Process tool results
 * - Handle model switching
 */
export class ContextManager {
  private model: string;
  private providerType: ProviderType;
  private contextLimit: number;
  private customSystemPrompt: string | undefined;
  private workingDirectory: string;
  private tools: Tool[];
  private messageStore: MessageStore;
  private compactionEngine: CompactionEngine;
  private snapshotBuilder: ContextSnapshotBuilder;
  /** Rules content from AGENTS.md / CLAUDE.md hierarchy */
  private rulesContent: string | undefined;

  // Cached values for performance
  private cachedSystemPromptTokens: number | null = null;
  private cachedToolsTokens: number | null = null;

  /**
   * API-reported context token count from model response.
   * This is the ground truth from the actual tokenizer, set after each turn.
   * When available, this is preferred over our char/4 estimates.
   * null = no API data available (use component estimates as fallback).
   */
  private lastApiContextTokens: number | null = null;

  constructor(config: ContextManagerConfig) {
    this.model = config.model;
    this.providerType = detectProviderFromModel(config.model);
    this.contextLimit = getContextLimit(config.model);
    this.workingDirectory = config.workingDirectory ?? process.cwd();
    this.tools = config.tools ?? [];
    this.rulesContent = config.rulesContent;
    this.messageStore = new MessageStore();

    // Create compaction engine with dependencies injected
    this.compactionEngine = new CompactionEngine(
      {
        threshold: config.compaction?.threshold ?? 0.70,
        preserveRecentTurns: config.compaction?.preserveRecentTurns ?? 5,
      },
      {
        getMessages: () => this.messageStore.get(),
        setMessages: (msgs) => this.setMessages(msgs),
        getCurrentTokens: () => this.getCurrentTokens(),
        getContextLimit: () => this.contextLimit,
        estimateSystemPromptTokens: () => this.estimateSystemPromptTokens(),
        estimateToolsTokens: () => this.estimateToolsTokens(),
        getMessageTokens: (msg) =>
          this.messageStore.getCachedTokens(msg) ?? estimateMessageTokensUtil(msg),
      }
    );

    // Create snapshot builder with dependencies injected
    this.snapshotBuilder = new ContextSnapshotBuilder({
      getCurrentTokens: () => this.getCurrentTokens(),
      getContextLimit: () => this.contextLimit,
      getMessages: () => this.messageStore.get(),
      estimateSystemPromptTokens: () => this.estimateSystemPromptTokens(),
      estimateToolsTokens: () => this.estimateToolsTokens(),
      estimateRulesTokens: () => this.estimateRulesTokens(),
      getMessagesTokens: () => this.getMessagesTokens(),
      getMessageTokens: (msg) =>
        this.messageStore.getCachedTokens(msg) ?? estimateMessageTokensUtil(msg),
      getSystemPrompt: () => this.getSystemPrompt(),
      getToolClarificationMessage: () => this.getToolClarificationMessage(),
      getTools: () => this.tools,
    });

    // Load system prompt with priority: programmatic > file-based > hardcoded
    // Note: Check for undefined, not truthiness, so empty string '' is honored
    if (config.systemPrompt !== undefined) {
      // Explicit programmatic override takes highest priority
      this.customSystemPrompt = config.systemPrompt;
    } else {
      // Try loading from project .tron/SYSTEM.md
      const loaded = loadSystemPromptFromFileSync({
        workingDirectory: this.workingDirectory,
      });
      if (loaded) {
        this.customSystemPrompt = loaded.content;
      } else {
        // No explicit prompt and no project SYSTEM.md - will use TRON_CORE_PROMPT
        this.customSystemPrompt = undefined;
      }
    }
  }

  // ===========================================================================
  // Message Management
  // ===========================================================================

  /**
   * Add a message to the context.
   */
  addMessage(message: Message): void {
    this.messageStore.add(message);
  }

  /**
   * Replace all messages.
   */
  setMessages(messages: Message[]): void {
    this.messageStore.set(messages);
    // Reset API tokens since message set changed — fall back to estimates until next turn
    this.lastApiContextTokens = null;
  }

  /**
   * Get all messages (defensive copy).
   */
  getMessages(): Message[] {
    return this.messageStore.get();
  }

  /**
   * Clear all messages from context.
   * System prompt and tools are preserved.
   */
  clearMessages(): void {
    this.messageStore.clear();
    // Reset API tokens — fall back to estimates until next turn completes
    this.lastApiContextTokens = null;
  }

  /**
   * Get the system prompt built for the current provider.
   * For most providers, this returns the Tron system prompt.
   * For OpenAI Codex, returns empty string (Tron context is via tool clarification).
   */
  getSystemPrompt(): string {
    return buildSystemPrompt({
      providerType: this.providerType,
      workingDirectory: this.workingDirectory,
      customPrompt: this.customSystemPrompt,
    });
  }

  /**
   * Get the raw/custom system prompt without provider adaptation.
   */
  getRawSystemPrompt(): string {
    return this.customSystemPrompt ?? TRON_CORE_PROMPT;
  }

  /**
   * Get tool clarification message for providers that need it (e.g., OpenAI Codex).
   * Returns null if the provider uses standard system prompts.
   */
  getToolClarificationMessage(): string | null {
    return getToolClarificationMessage({
      providerType: this.providerType,
      workingDirectory: this.workingDirectory,
      customPrompt: this.customSystemPrompt,
    });
  }

  /**
   * Check if current provider requires tool clarification instead of system prompt.
   */
  requiresToolClarification(): boolean {
    return requiresToolClarificationMessage(this.providerType);
  }

  /**
   * Get current provider type.
   */
  getProviderType(): ProviderType {
    return this.providerType;
  }

  /**
   * Get working directory.
   */
  getWorkingDirectory(): string {
    return this.workingDirectory;
  }

  /**
   * Set rules content from AGENTS.md / CLAUDE.md hierarchy.
   * Invalidates cached token counts.
   */
  setRulesContent(content: string | undefined): void {
    this.rulesContent = content;
    this.invalidateSystemPromptCache();
  }

  /**
   * Get rules content from AGENTS.md / CLAUDE.md hierarchy.
   */
  getRulesContent(): string | undefined {
    return this.rulesContent;
  }

  /**
   * Set working directory (updates cached system prompt tokens).
   */
  setWorkingDirectory(dir: string): void {
    this.workingDirectory = dir;
    this.invalidateSystemPromptCache();
  }

  /**
   * Get tools.
   */
  getTools(): Tool[] {
    return [...this.tools];
  }

  // ===========================================================================
  // Token Tracking
  // ===========================================================================

  /**
   * Get current total token count.
   * Returns API-reported tokens when available (ground truth from model's tokenizer).
   * Falls back to sum of component estimates when API data is unavailable
   * (before first turn, after compaction resets messages, after clearMessages).
   */
  getCurrentTokens(): number {
    if (this.lastApiContextTokens !== null) {
      return this.lastApiContextTokens;
    }
    return this.estimateSystemPromptTokens()
      + this.estimateToolsTokens()
      + this.estimateRulesTokens()
      + this.getMessagesTokens();
  }

  /**
   * Get model's context limit.
   */
  getContextLimit(): number {
    return this.contextLimit;
  }

  /**
   * Get current model.
   */
  getModel(): string {
    return this.model;
  }

  /**
   * Set the API-reported context token count.
   * Called after each turn with actual tokenizer count from model API.
   * This value is preferred over estimates when available.
   */
  setApiContextTokens(tokens: number): void {
    this.lastApiContextTokens = tokens;
  }

  /**
   * Get the API-reported context token count (0 if not yet available).
   */
  getApiContextTokens(): number {
    return this.lastApiContextTokens ?? 0;
  }

  // ===========================================================================
  // Snapshot & Validation
  // ===========================================================================

  /**
   * Get a snapshot of current context state.
   * For UI consistency, uses API tokens when available (matches progress bar).
   * Before first turn, shows 0 tokens used (full context remaining).
   */
  getSnapshot(): ContextSnapshot {
    return this.snapshotBuilder.build();
  }

  /**
   * Get a detailed snapshot with per-message token breakdown.
   */
  getDetailedSnapshot(): DetailedContextSnapshot {
    return this.snapshotBuilder.buildDetailed();
  }

  /**
   * Check if a new turn can be accepted.
   */
  canAcceptTurn(opts: { estimatedResponseTokens: number }): PreTurnValidation {
    const currentTokens = this.getCurrentTokens();
    const estimated = currentTokens + opts.estimatedResponseTokens;
    const level = this.getThresholdLevel(currentTokens);

    return {
      canProceed: level !== 'critical' && level !== 'exceeded',
      needsCompaction: level === 'alert' || level === 'critical' || level === 'exceeded',
      wouldExceedLimit: estimated > this.contextLimit,
      currentTokens,
      estimatedAfterTurn: estimated,
      contextLimit: this.contextLimit,
    };
  }

  /**
   * Get threshold level for a token count.
   */
  getThresholdLevel(tokens: number): ThresholdLevel {
    return this.snapshotBuilder.getThresholdLevel(tokens);
  }

  // ===========================================================================
  // Model Switching
  // ===========================================================================

  /**
   * Switch to a different model.
   * Updates provider type, revalidates context, and triggers callback if compaction is needed.
   */
  switchModel(newModel: string): void {
    this.model = newModel;
    this.providerType = detectProviderFromModel(newModel);
    this.contextLimit = getContextLimit(newModel);
    this.invalidateSystemPromptCache();

    // Check if we now need compaction
    this.compactionEngine.triggerIfNeeded();
  }

  /**
   * Register callback for when compaction is needed.
   */
  onCompactionNeeded(callback: () => void): void {
    this.compactionEngine.onNeeded(callback);
  }

  // ===========================================================================
  // Compaction
  // ===========================================================================

  /**
   * Check if compaction is recommended.
   */
  shouldCompact(): boolean {
    return this.compactionEngine.shouldCompact();
  }

  /**
   * Generate a compaction preview without modifying state.
   */
  async previewCompaction(opts: {
    summarizer: Summarizer;
  }): Promise<CompactionPreview> {
    return this.compactionEngine.preview(opts.summarizer);
  }

  /**
   * Execute compaction and update messages.
   */
  async executeCompaction(opts: {
    summarizer: Summarizer;
    editedSummary?: string;
  }): Promise<CompactionResult> {
    return this.compactionEngine.execute(opts);
  }

  // ===========================================================================
  // Tool Result Processing
  // ===========================================================================

  /**
   * Process a tool result, potentially truncating if too large.
   */
  processToolResult(result: {
    toolCallId: string;
    content: string;
  }): ProcessedToolResult {
    const maxSize = this.getMaxToolResultSize();
    if (result.content.length <= maxSize) {
      return { ...result, truncated: false };
    }

    return {
      toolCallId: result.toolCallId,
      content: result.content.slice(0, maxSize) + '\n... [truncated]',
      truncated: true,
      originalSize: result.content.length,
    };
  }

  /**
   * Get maximum tool result size based on remaining context budget.
   * This provides a dynamic ceiling that adapts to current context usage.
   */
  getMaxToolResultSize(): number {
    const currentTokens = this.getCurrentTokens();
    const remainingBudget = this.contextLimit - currentTokens;

    // Reserve tokens for:
    // - Model response: 8,000 tokens (typical response size)
    // - Safety margin: 10% of remaining budget
    const responseReserve = 8_000;
    const safetyMargin = Math.floor(remainingBudget * 0.1);

    const availableTokens = Math.max(
      remainingBudget - responseReserve - safetyMargin,
      2_500  // Minimum 2,500 tokens for tool results
    );

    // Convert to chars (4 chars ~ 1 token), cap at 100k chars (~25k tokens)
    return Math.min(availableTokens * 4, 100_000);
  }

  // ===========================================================================
  // Serialization
  // ===========================================================================

  /**
   * Export state for persistence.
   */
  exportState(): ExportedState {
    return {
      model: this.model,
      systemPrompt: this.getSystemPrompt(),
      tools: [...this.tools],
      messages: this.messageStore.get(),
    };
  }

  // ===========================================================================
  // Private Helpers
  // ===========================================================================

  /**
   * Invalidate cached system prompt tokens.
   * Called when rules content, working directory, or model changes.
   */
  private invalidateSystemPromptCache(): void {
    this.cachedSystemPromptTokens = null;
  }

  private estimateSystemPromptTokens(): number {
    if (this.cachedSystemPromptTokens !== null) {
      return this.cachedSystemPromptTokens;
    }
    this.cachedSystemPromptTokens = estimateSystemPromptTokensUtil(
      this.getSystemPrompt(),
      this.getToolClarificationMessage()
    );
    return this.cachedSystemPromptTokens;
  }

  private estimateToolsTokens(): number {
    if (this.cachedToolsTokens !== null) {
      return this.cachedToolsTokens;
    }
    this.cachedToolsTokens = estimateToolsTokensUtil(this.tools);
    return this.cachedToolsTokens;
  }

  private estimateRulesTokens(): number {
    return estimateRulesTokensUtil(this.rulesContent);
  }

  private getMessagesTokens(): number {
    return this.messageStore.getTokens();
  }
}

// =============================================================================
// Factory
// =============================================================================

/**
 * Create a new ContextManager instance.
 */
export function createContextManager(config: ContextManagerConfig): ContextManager {
  return new ContextManager(config);
}

// =============================================================================
// Type Re-exports (public API)
// =============================================================================

export type {
  ThresholdLevel,
  ContextManagerConfig,
  RulesFileSnapshot,
  RulesSnapshot,
  ContextSnapshot,
  DetailedMessageInfo,
  DetailedContextSnapshot,
  PreTurnValidation,
  CompactionPreview,
  CompactionResult,
  ProcessedToolResult,
  ExportedState,
} from './types.js';

File: packages/agent/src/domains/agent/runner/context/context_manager.rs
//! Primitive context manager.
//!
//! The manager owns only loop infrastructure: message history, the audited soul
//! prompt, provider-visible capability schemas for accounting, environment
//! metadata, token estimates, and compaction state.

use std::sync::Arc;

use crate::shared::messages::Message;

use super::compaction_engine::{CompactionDeps, CompactionEngine};
use super::constants::{
    CAPABILITY_RESULT_MAX_CHARS, CAPABILITY_RESULT_MIN_TOKENS, CHARS_PER_TOKEN, Thresholds,
};
use super::context_snapshot_builder::ContextSnapshotBuilder;
use super::message_store::MessageStore;
use super::summarizer::Summarizer;
use super::token_estimator;
use super::types::{
    CompactionPreview, CompactionResult, ContextManagerConfig, ContextSnapshot,
    DetailedContextSnapshot, ExportedState, ExtractedData, PreTurnValidation,
    ProcessedCapabilityResult,
};

mod compaction_deps;
mod snapshot_deps;

use compaction_deps::ManagerCompactionDeps;
use snapshot_deps::ManagerSnapshotDeps;

pub struct ContextManager {
    pub(super) config: ContextManagerConfig,
    messages: MessageStore,
    api_context_tokens: Option<u64>,
    system_prompt: String,
    last_extracted_data: Option<ExtractedData>,
    server_origin: Option<String>,
    turn_generation: u64,
    turn_shape_refreshed_at_generation: Option<u64>,
}

impl ContextManager {
    pub fn new(mut config: ContextManagerConfig) -> Self {
        if config.working_directory.is_none() {
            let home = crate::shared::paths::home_dir();
            config.working_directory = Some(format!("{home}/Workspace"));
        }

        let system_prompt = config.system_prompt.clone().unwrap_or_else(|| {
            panic!("ContextManagerConfig.system_prompt must be resolved before construction")
        });

        Self {
            config,
            messages: MessageStore::new(),
            api_context_tokens: None,
            system_prompt,
            last_extracted_data: None,
            server_origin: None,
            turn_generation: 0,
            turn_shape_refreshed_at_generation: None,
        }
    }

    pub fn begin_turn(&mut self) {
        self.turn_generation = self.turn_generation.saturating_add(1);
    }

    #[must_use]
    pub fn turn_generation(&self) -> u64 {
        self.turn_generation
    }

    #[must_use]
    pub fn volatile_tokens_fresh_for_current_turn(&self) -> bool {
        self.turn_shape_refreshed_at_generation == Some(self.turn_generation)
    }

    pub fn add_message(&mut self, message: Message) {
        self.messages.add(message);
    }

    pub fn set_messages(&mut self, messages: Vec<Message>) {
        self.messages.set(messages);
        self.api_context_tokens = None;
    }

    #[must_use]
    pub fn get_messages(&self) -> Vec<Message> {
        self.messages.as_slice().to_vec()
    }

    pub fn get_messages_arc(&mut self) -> Arc<[Message]> {
        self.messages.as_arc()
    }

    #[must_use]
    pub fn messages_slice(&self) -> &[Message] {
        self.messages.as_slice()
    }

    pub fn clear_messages(&mut self) {
        self.messages.clear();
        self.api_context_tokens = None;
    }

    #[must_use]
    pub fn message_count(&self) -> usize {
        self.messages.len()
    }

    #[must_use]
    pub fn get_system_prompt(&self) -> &str {
        &self.system_prompt
    }

    #[must_use]
    pub fn get_current_tokens(&self) -> u64 {
        if let Some(api_tokens) = self.api_context_tokens {
            return api_tokens;
        }
        self.estimate_system_prompt_tokens()
            + self.estimate_capabilities_tokens()
            + self.estimate_environment_tokens()
            + self.get_messages_tokens()
    }

    pub fn set_api_context_tokens(&mut self, tokens: u64) {
        self.api_context_tokens = Some(tokens);
    }

    #[must_use]
    pub fn get_api_context_tokens(&self) -> Option<u64> {
        self.api_context_tokens
    }

    #[must_use]
    pub fn get_context_limit(&self) -> u64 {
        self.config.compaction.context_limit
    }

    #[must_use]
    pub fn get_model(&self) -> &str {
        &self.config.model
    }

    #[must_use]
    pub fn get_working_directory(&self) -> &str {
        self.config.working_directory.as_deref().unwrap_or("/tmp")
    }

    #[must_use]
    pub fn estimate_system_prompt_tokens(&self) -> u64 {
        u64::from(token_estimator::estimate_system_prompt_tokens(
            &self.system_prompt,
            None,
        ))
    }

    #[must_use]
    pub fn estimate_capabilities_tokens(&self) -> u64 {
        u64::from(token_estimator::estimate_capabilities_tokens(
            &self.config.capabilities,
        ))
    }

    #[must_use]
    pub fn model_capability_names(&self) -> Vec<String> {
        self.config
            .capabilities
            .iter()
            .map(|capability| capability.name.clone())
            .collect()
    }

    #[must_use]
    pub fn get_messages_tokens(&self) -> u64 {
        u64::from(self.messages.get_tokens())
    }

    #[must_use]
    pub fn get_message_tokens(&self, msg: &Message) -> u64 {
        u64::from(token_estimator::estimate_message_tokens(msg))
    }

    #[must_use]
    pub fn estimate_environment_tokens(&self) -> u64 {
        let wd = self
            .config
            .working_directory
            .as_ref()
            .map_or(0, |wd| (wd.len() + 30) as u64 / CHARS_PER_TOKEN as u64);
        let origin = self.server_origin.as_ref().map_or(0, |origin| {
            (origin.len() + 10) as u64 / CHARS_PER_TOKEN as u64
        });
        wd + origin
    }

    pub fn set_server_origin(&mut self, origin: Option<String>) {
        self.server_origin = origin;
    }

    pub fn set_volatile_tokens(&mut self, _a: u64, _b: u64, _c: u64) {
        self.turn_shape_refreshed_at_generation = Some(self.turn_generation);
    }

    #[must_use]
    pub fn get_snapshot(&self) -> ContextSnapshot {
        self.assert_turn_shape_refreshed("get_snapshot");
        ContextSnapshotBuilder::new(ManagerSnapshotDeps { manager: self }).build()
    }

    #[must_use]
    pub fn get_detailed_snapshot(&self) -> DetailedContextSnapshot {
        self.assert_turn_shape_refreshed("get_detailed_snapshot");
        ContextSnapshotBuilder::new(ManagerSnapshotDeps { manager: self }).build_detailed()
    }

    fn assert_turn_shape_refreshed(&self, caller: &'static str) {
        if self.turn_generation == 0 {
            return;
        }
        if !self.volatile_tokens_fresh_for_current_turn() {
            debug_assert!(
                false,
                "turn context shape not refreshed for generation {} before {}",
                self.turn_generation, caller
            );
            tracing::warn!(
                generation = self.turn_generation,
                recorded = ?self.turn_shape_refreshed_at_generation,
                caller,
                "turn context shape was stale for current turn"
            );
        }
    }

    #[must_use]
    pub fn can_accept_turn(&self) -> PreTurnValidation {
        let current = self.get_current_tokens();
        let limit = self.get_context_limit();
        #[allow(clippy::cast_precision_loss)]
        let ratio = if limit > 0 {
            current as f64 / limit as f64
        } else {
            0.0
        };
        PreTurnValidation {
            can_proceed: ratio < Thresholds::CRITICAL,
            needs_compaction: ratio >= self.config.compaction.threshold,
            current_tokens: current,
            context_limit: limit,
        }
    }

    #[must_use]
    pub fn should_compact(&self) -> bool {
        let limit = self.get_context_limit();
        if limit == 0 {
            return false;
        }
        #[allow(clippy::cast_precision_loss)]
        let ratio = self.get_current_tokens() as f64 / limit as f64;
        ratio >= self.config.compaction.threshold
    }

    #[must_use]
    pub fn has_summarizable_compaction_window(&self) -> bool {
        let deps = ManagerCompactionDeps::from_manager(self);
        let engine = CompactionEngine::new(
            self.config.compaction.threshold,
            self.config.compaction.preserve_recent_turns,
            deps,
        );
        engine.has_summarizable_messages()
    }

    pub async fn preview_compaction(
        &self,
        summarizer: &dyn Summarizer,
    ) -> Result<CompactionPreview, Box<dyn std::error::Error + Send + Sync>> {
        let deps = ManagerCompactionDeps::from_manager(self);
        let engine = CompactionEngine::new(
            self.config.compaction.threshold,
            self.config.compaction.preserve_recent_turns,
            deps,
        );
        engine.preview(summarizer).await
    }

    pub async fn execute_compaction(
        &mut self,
        summarizer: &dyn Summarizer,
        edited_summary: Option<&str>,
    ) -> Result<CompactionResult, Box<dyn std::error::Error + Send + Sync>> {
        let deps = ManagerCompactionDeps::from_manager(self);
        let engine = CompactionEngine::new(
            self.config.compaction.threshold,
            self.config.compaction.preserve_recent_turns,
            deps,
        );
        let result = engine.execute(summarizer, edited_summary).await?;
        self.messages.set(engine.deps.get_messages());
        if let Some(ref data) = result.extracted_data {
            self.last_extracted_data = Some(data.clone());
        }
        self.api_context_tokens = None;
        Ok(result)
    }

    #[must_use]
    pub fn process_capability_result(
        &self,
        invocation_id: &str,
        content: &str,
    ) -> ProcessedCapabilityResult {
        let max_size = self.get_max_capability_result_size();
        if content.len() <= max_size {
            ProcessedCapabilityResult {
                invocation_id: invocation_id.to_owned(),
                content: content.to_owned(),
                truncated: false,
                original_size: None,
            }
        } else {
            let body_budget = max_size.saturating_sub(100);
            let prefix = crate::shared::text::truncate_str(content, body_budget);
            ProcessedCapabilityResult {
                invocation_id: invocation_id.to_owned(),
                content: format!(
                    "{prefix}...\n[Truncated: {} chars total, showing first {}]",
                    content.len(),
                    prefix.len()
                ),
                truncated: true,
                original_size: Some(content.len()),
            }
        }
    }

    #[must_use]
    pub fn get_max_capability_result_size(&self) -> usize {
        let limit = self.get_context_limit();
        let current = self.get_current_tokens();
        let remaining = limit.saturating_sub(current);
        let response_reserve: u64 = 8_000;
        let margin: u64 = remaining / 10;
        let available_tokens = remaining
            .saturating_sub(response_reserve)
            .saturating_sub(margin)
            .max(u64::from(CAPABILITY_RESULT_MIN_TOKENS));

        #[allow(clippy::cast_possible_truncation)]
        let budget = (available_tokens as usize) * (CHARS_PER_TOKEN as usize);
        budget.min(CAPABILITY_RESULT_MAX_CHARS)
    }

    pub fn switch_model(&mut self, new_model: String, context_limit: u64) {
        self.config.model = new_model;
        self.config.compaction.context_limit = context_limit;
        self.api_context_tokens = None;
    }

    #[must_use]
    pub fn build_base_context(&self) -> crate::shared::messages::Context {
        crate::shared::messages::Context {
            system_prompt: Some(self.get_system_prompt().to_owned()),
            messages: Arc::default(),
            capabilities: None,
            working_directory: Some(self.get_working_directory().to_owned()),
            agent_state_context: None,
            server_origin: None,
        }
    }

    #[must_use]
    pub fn get_latest_extracted_data(&self) -> ExtractedData {
        self.last_extracted_data.clone().unwrap_or_default()
    }

    pub fn set_extracted_data(&mut self, data: ExtractedData) {
        self.last_extracted_data = Some(data);
    }

    #[must_use]
    pub fn export_state(&self) -> ExportedState {
        ExportedState {
            model: self.config.model.clone(),
            system_prompt: self.system_prompt.clone(),
            capabilities: self.config.capabilities.clone(),
            messages: self.get_messages(),
        }
    }
}

#[cfg(test)]
#[path = "context_manager_tests.rs"]
mod tests;
